add credential format specific sections for IAR endpoint binding in VPs (#602)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

The Wallet MUST include the most recently received `auth_session` in follow-up requests to the Interactive Authorization Endpoint.

If a Wallet receives a `type` value that it does not recognize, it MUST abort the issuance process.

Additional keys are defined based on the type of interaction, as shown next.

#### Require Presentation {#iae-require-presentation}

If `type` is set to `openid4vp_presentation`, as shown in the following example, the response MUST further include an `openid4vp_request` parameter containing an OpenID4VP Authorization Request. The contents of the request is the same as for requests passed to the Digital Credentials API (see Appendix A.2 and Appendix A.3 of [@!OpenID4VP]), except as follows: 

> * The `response_mode` must be either `iae-post` for unencrypted responses or `iae-post.jwt` for encrypted responses. These modes are used to indicate to the Wallet to return the response back to the same Interactive Authorization Endpoint. 

Reviewer: ```suggestion
* The `response_mode` must be either `iae_post` for unencrypted responses or `iae_post.jwt` for encrypted responses. These modes are used to indicate to the Wallet to return the response back to the same Interactive Authorization Endpoint. 
```

Nit, but why not keep it consistent with direct_post casing?

Author: Ok, updated the PR.

When processing the request the following logic applies:

>   1. The URL of the Interactive Authorization Endpoint becomes the Origin for the request; i.e., the Wallet MUST ensure that `expected_origins` contains the Interactive Authorization Endpoint URL.

Reviewer 2: Two comments:
1. since expected_origins is optional (or conditional), should we add "if present"?
2. expected_origins can only contain origins, not URLs. While they can match, they don't always match. Is the idea that the expected_origins is converted by the wallet before matching?

Author: I agree that the current definition of `Origin` and `expected_origins` in OID4VP clashes with this normative statement. `expected_origins` is conditional based on if the request was signed. If somebody wants to write a parser for requests, this would be prone to errors since with the current text both, URLs and origins, are allowed.

I also agree that it would be cleaner to say "if present" although I don't know whether this was intentional and part of the original threat mitigation model specifically for IAE requests. I believe it wasn't, so I'm fine with adding "if present".

What we could do (non-breaking):
1. Keep using `expected_origins` and use the origin part of the URL only. Probably has some security implications.
2. same as 1. (since required for signed requests) but additionally define that the URL needs to be checked against the `interactive_authorization_endpoint` from the metadata, or define a new optional parameter for the request for signed requests with similar logic to `expected_origins`, i.e., `expected_urls`.

>   2. If the response includes Verifiable Presentations with Holder Binding, each such Verifiable Presentation MUST be bound to the Interactive Authorization Endpoint, as specified by the corresponding Credential Format. For Credential Format-specific binding mechanisms, see "Interactive Authorization Endpoint Binding" subsections in (#format-profiles). In this case, the Credential Format refers to the Credential Format of the Verifiable Presentation requested by the OpenID4VP Authorization Request, which can differ from the Credential Format of the Credentials being issued. A response MUST be rejected if any such Verifiable Presentation is not bound to the Interactive Authorization Endpoint accordingly.

Reviewer 3: ```suggestion
  2. If the response contains Verifiable Presentations that include Holder Binding, each of those must be properly bound to the Interactive Authorization Endpoint, following the rules defined by their Credential Format. Details on how to do this for each format can be found in the "Interactive Authorization Endpoint Binding" sections under (#format-profiles). Note that the Credential Format here refers to the format of the Verifiable Presentation requested in the OpenID4VP Authorization Request, which may be different from the format used for issuing the Credentials themselves. If any Verifiable Presentation with Holder Binding is not correctly bound to the Interactive Authorization Endpoint, the response MUST be rejected.
```

When processing the request the following logic applies:

  1. If `expected_origins` is present, the Wallet MUST ensure that `expected_origins` contains the derived Origin as defined above.

>   2. If the response contains Verifiable Presentations that include Holder Binding, each of those MUST be properly bound to the Interactive Authorization Endpoint, following the rules defined by their Credential Format. Details on how to do this for each format can be found in the "Interactive Authorization Endpoint Binding" sections under (#format-profiles). Note that the Credential Format here refers to the format of the Verifiable Presentation requested in the OpenID4VP Authorization Request, which may be different from the format used for issuing the Credentials themselves. If any Verifiable Presentation with Holder Binding is not correctly bound to the Interactive Authorization Endpoint, the response MUST be rejected.

Author: This is more cohesive with OID4VP
```suggestion
  2. If the response contains Verifiable Presentations that include Holder Binding, the audience of each of those MUST be properly bound to the Interactive Authorization Endpoint, following the rules defined by their Credential Format. Details on how to do this for each format can be found in the "Interactive Authorization Endpoint Binding" sections under (#format-profiles). Note that the Credential Format here refers to the format of the Verifiable Presentation requested in the OpenID4VP Authorization Request, which may be different from the format used for issuing the Credentials themselves. If any Verifiable Presentation with Holder Binding is not correctly bound to the Interactive Authorization Endpoint, the response MUST be rejected.
```

<{{examples/credential_response_mso_mdoc.txt}}

> ### Interactive Authorization Endpoint Binding {#iae-binding-mso-mdoc}

Author: @Sakurann Yep, makes sense. How about the suggestion below?

* The second element MUST be a Byte String which contains the sha-256 hash of the bytes of `OpenID4VCIIAEHandoverInfo` when encoded as CBOR.
* The `OpenID4VCIIAEHandoverInfo` has the following elements:
  * The first element MUST be the string representing the Interactive Authorization Endpoint URL of the request as described in (#interactive-authorization-request). It MUST NOT be prefixed with `iar:`.
  * The second element MUST be the value of the `nonce` request parameter.
  * For the Response Mode `iar-post.jwt`, the third element MUST be the JWK SHA-256 Thumbprint as defined in [@!RFC7638], encoded as a Byte String, of the Verifier's public key used to encrypt the response. If the Response Mode is `iar-post`, the third element MUST be `null`. For unsigned requests, including the JWK Thumbprint in the `SessionTranscript` allows the Verifier to detect whether the response was re-encrypted by a third party, potentially leading to the leakage of sensitive information. While this does not prevent such an attack, it makes it detectable and helps preserve the confidentiality of the response.  

> The following is a non-normative example of the input JWK for calculating the JWK Thumbprint in the context of `OpenID4VCIIAEHandoverInfo`:

Reviewer 3: ```suggestion
The following is a non-normative example of the input JWK for calculating the JWK Thumbprint in the context of `OpenID4VCIIAEHandoverInfo`:

```

Author: It compiled for me locally without the extra lines but I can certainly commit your suggestion.

```json
{
  "kty": "EC",
  "crv": "P-256",
  "x": "DxiH5Q4Yx3UrukE2lWCErq8N8bqC9CHLLrAwLz5BmE0",
  "y": "XtLM4-3h5o3HUH0MHVJV0kyq0iBlrBwlh8qEDMZ4-Pc",
  "use": "enc",
  "alg": "ECDH-ES",
  "kid": "1"
}
```

> The following is a non-normative example of the `OpenID4VCIIAEHandoverInfo` structure:

Reviewer 3: ```suggestion
The following is a non-normative example of the `OpenID4VCIIAEHandoverInfo` structure:

```

```
Hex:

837768747470733a2f2f6578616d706c652e636f6d2f696172782b6578633767426b
786a7831726463397564527276654b7653734a4971383061766c58654c4868477771
744158204283ec927ae0f208daaa2d026a814f2b22dca52cf85ffa8f3f8626c6bd66
9047

CBOR diagnostic:

83                                   # array(3)
  77                                 #   string(23)
    68747470733a2f2f6578616d706c652e #     "https://example."
    636f6d2f696172                   #     "com/iar"
  78 2b                              #   string(43)
    6578633767426b786a78317264633975 #     "exc7gBkxjx1rdc9u"
    64527276654b7653734a497138306176 #     "dRrveKvSsJIq80av"
    6c58654c48684777717441           #     "lXeLHhGwqtA"
  58 20                              #   bytes(32)
    4283ec927ae0f208daaa2d026a814f2b #     "B\x83ì\x92zàò\x08Úª-\x02j\x81O+"
    22dca52cf85ffa8f3f8626c6bd669047 #     ""Ü¥,ø_ú\x8f?\x86&Æ½f\x90G"
```

> The following is a non-normative example of the `OpenID4VCIIAEHandover` structure:

Reviewer 3: ```suggestion
The following is a non-normative example of the `OpenID4VCIIAEHandover` structure:

```

```
Hex:

82754f70656e49443456434949414548616e646f7665725820cf3cbd10e9d68754ef
01c0fa0fab3188a757bdab209a2c4b6235f1597b9d2e16

CBOR diagnostic:

82                                   # array(2)
  75                                 #   string(21)
    4f70656e49443456434949414548616e #     "OpenID4VCIIAEHan"
    646f766572                       #     "dover"
  58 20                              #   bytes(32)
    cf3cbd10e9d68754ef01c0fa0fab3188 #     "Ï<½\x10éÖ\x87Tï\x01Àú\x0f«1\x88"
    a757bdab209a2c4b6235f1597b9d2e16 #     "§W½« \x9a,Kb5ñY{\x9d.\x16"
```

> The following is a non-normative example of the `SessionTranscript` structure:

Reviewer 3: ```suggestion
The following is a non-normative example of the `SessionTranscript` structure:

```

<{{examples/credential_response_mso_mdoc.txt}}

### Interactive Authorization Endpoint Binding {#iae-binding-mso-mdoc}

> To bind the Interactive Authorization Endpoint to a Verifiable Presentation using the Credential Format defined in this section, the `SessionTranscript` CBOR structured as defined in Section 9.1.5.1 in [@ISO.18013-5] MUST be used with the following changes:

Author: @Sakurann is this better?
```suggestion
To bind the Interactive Authorization Endpoint to a Verifiable Presentation using the Credential Format defined in this section, the `SessionTranscript` CBOR structured as defined in Section 9.1.5.1 in [@ISO.18013-5] MUST be used in Verifiable Presentations submitted in a response to Interactive Authorization Requests using the `openid4vp_presentation` interaction type, with the following modifications. This `SessionTranscript` differs from those defined in Section B.5.6 in [@OpenID4VP] and is defined as follows:
```

Reviewer 3: yes!

* `DeviceEngagementBytes` MUST be `null`.
* `EReaderKeyBytes` MUST be `null`.
* `Handover` MUST be the `OpenID4VCIIAEHandover` CBOR structure as defined below.

Note: The following section contains a definition in Concise Data Definition Language (CDDL), a language used to define data structures - see [@RFC8610] for more details. `bstr` refers to Byte String, defined as major type 2 in CBOR and `tstr` refers to Text String, defined as major type 3 in CBOR (encoded in utf-8) as defined in section 3.1 of [@RFC8949].

```cddl
OpenID4VCIIAEHandover = [
  "OpenID4VCIIAEHandover", ; A fixed identifier for this handover type
  OpenID4VCIIAEHandoverInfoHash ; A cryptographic hash of OpenID4VCIIAEHandoverInfo
]

; Contains the sha-256 hash of OpenID4VCIIAEHandoverInfoBytes
OpenID4VCIIAEHandoverInfoHash = bstr

; Contains the bytes of OpenID4VCIIAEHandoverInfo encoded as CBOR
OpenID4VCIIAEHandoverBytes = bstr .cbor OpenID4VCIIAEHandoverInfo

OpenID4VCIIAEHandoverInfo = [

>   iaeUrl,

Author: We didn't prefix with `origin:` in DC API either. My thought was that the type in the `SessionTranscript` would replace the prefixing and this is the way we provide the context on the flow for session transcripts across flows.
